src/main/java/com/azure/cosmos/examples/bulk/sync/BulkWriter.java

```diff
@@ -16,134 +16,159 @@
 import com.azure.cosmos.models.CosmosItemOperation;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
-import reactor.core.publisher.Sinks;
+
+import java.util.ArrayList;
+import java.util.List;
 import java.util.concurrent.Semaphore;
 
 public class BulkWriter {
     private static final Logger logger = LoggerFactory.getLogger(BulkWriter.class);
 
-    private final Sinks.Many<CosmosItemOperation> bulkInputEmitter = Sinks.many().unicast().onBackpressureBuffer();
+    private final CosmosContainer cosmosContainer;
     private final int cpuCount = Runtime.getRuntime().availableProcessors();
-
-    //Max items to be buffered to avoid out of memory error
     private final Semaphore semaphore = new Semaphore(1024 * 167 / cpuCount);
-
-    private final Sinks.EmitFailureHandler emitFailureHandler =
-            (signalType, emitResult) -> {
-                if (emitResult.equals(Sinks.EmitResult.FAIL_NON_SERIALIZED)) {
-                    logger.debug("emitFailureHandler - Signal: [{}], Result: [{}]", signalType, emitResult);
-                    return true;
-                } else {
-                    logger.error("emitFailureHandler - Signal: [{}], Result: [{}]", signalType, emitResult);
-                    return false;
-                }
-            };
-
-    private final CosmosContainer cosmosContainer;
+    private final List<CosmosItemOperation> bufferedOperations = new ArrayList<>();
+    private final int maxRetries = 5;
 
     public BulkWriter(CosmosContainer cosmosContainer) {
         this.cosmosContainer = cosmosContainer;
     }
 
     public void scheduleWrites(CosmosItemOperation cosmosItemOperation) {
-        while(!semaphore.tryAcquire()) {
+        while (!semaphore.tryAcquire()) {
             logger.info("Unable to acquire permit");
         }
         logger.info("Acquired permit");
-        scheduleInternalWrites(cosmosItemOperation);
-    }
-
-    private void scheduleInternalWrites(CosmosItemOperation cosmosItemOperation) {
-        bulkInputEmitter.emitNext(cosmosItemOperation, emitFailureHandler);
+        bufferedOperations.add(cosmosItemOperation);
```

Comment:
I think the sync implementation needs a broader conversation - not fully sure of the best practices with respect to buffering. In this specific implementation, `bufferedOperations` can be updated by multiple threads in parallel and isn't thread-safe and can lead to lost updates (semaphore can give away multiple permits).

Reply:
Will update to use `ConcurrentLinkedQueue`.